File: MultiChannelFxLMS.m
%% McFxLMS with arbitrary channel
% (J x K x M)
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%
%  Structure of McFxLMS
%
%                 M x J
%              +-----------+                       +   
% x(n) ---+--->|   P(z)    |--d(n)----------------> sum --+---> e(n)
%      J  |    +-----------+                          ^-   |  M
%         |                                           |    |
%         |        \ K x J                M x K     y(n)   |     
%         |    +-----------+    K     +-----------+   |    |
%         +--->|   W(z)    |--yc(n)-->|   S(z)    |---+    |
%         |    +-----------+          +-----------+        |
%         |            \                                   |
%         |             \----------------\                 |
%         |       M x K                   \                |
%         |    +-----------+          +-----------+        |
%         +--->|  Est_S(z) |--xf(n)-->|   LMS     |<-------+
%              +-----------+          +-----------+        
% 
%       x(n):  reference siganl/noise source
%       d(n):  disturbance signal
%       xf(n): filtered refence signal
%       yc(n): control signal
%       y(n):  anti noise
%       e(n):  error signal
%       P(Z):  Primary path
%       W(Z):  control filters
%       Est_S(Z): estimated Secondary path
%       S(Z):  Secondary path (real)
%      
%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%


classdef MultiChannelFxLMS

    properties
        Wc       % control filters K x J x wlen
        wlen     % length of control filter
        SecP     % secondary path M x K x slen
        slen     % length of secondary path
        Ref      % reference signal J x T
        Dis      % disturbance signal M x T
        yc       % control signal K x T
        y        % antinoise M x T
        Err      % error signal M x T
        J        % number of reference
        K        % number of secondary source
        M        % number of error
    end

    methods
        % initialization
        function obj =  MultiChannelFxLMS(wLen,Secondarypath,sLen,ref,dis,numref,numsec,numerr)
            T = size(ref,2);     % duration
            obj.J = numref;
            obj.K = numsec;
            obj.M = numerr;
            obj.Wc = zeros(numsec,numref,wLen);
            obj.wlen = wLen;
            obj.SecP = Secondarypath;
            obj.slen = sLen;
            obj.Ref = ref;
            obj.Dis = dis;
            obj.yc = zeros(numsec,T);
            obj.y = zeros(numerr,T);
            obj.Err = zeros(numerr,T);

        end

        % McFxLMS
        function [obj] = McFxLMS_controller(obj,muw)
            T = size(obj.Ref,2);

            xc = zeros(obj.J,obj.wlen);       % reference buffer for control filter
            ys = zeros(obj.K,obj.slen);       % control signal buffer for secondary path
            xf = zeros(obj.J,obj.slen);       % reference buffer for secondary path
            Xsf = zeros(obj.J,obj.K,obj.M,obj.wlen);  % filtered reference buffer
            
            for n = 1:T
                % reference buffer update
                xc = [obj.Ref(:,n) xc(:,1:end-1)];

                % control signal
                for kk = 1:obj.K
                    bb = reshape(obj.Wc(kk,:,:),[obj.J,obj.wlen]);
                    obj.yc(kk,n) = sum(sum(bb .* xc,2),1);
                end

                % anti noise
                ys = [obj.yc(:,n) ys(:,1:end-1)];
                for mm = 1:obj.M
                    cc = reshape(obj.SecP(mm,:,:),[obj.K,obj.slen]);
                    obj.y(mm,n) = sum(sum(cc .* ys,2),1);
                end

                % error
                obj.Err(:,n) = obj.Dis(:,n) - obj.y(:,n);

                % filtered reference
                xf = [obj.Ref(:,n) xf(:,1:end-1)];
                Xt = reshape(repmat(xf,(obj.K*obj.M),1),[obj.J obj.K obj.M obj.slen]);

                res = zeros(obj.J,obj.K,obj.M);
                for jj = 1:obj.J
                    dd = reshape(Xt(jj,:,:,:),[obj.K obj.M obj.slen]);
                    res(jj,:,:) = sum(dd .* permute(obj.SecP,[2 1 3]),3);
                end
                Xsf(:,:,:,2:end) = Xsf(:,:,:,1:end-1);
                Xsf(:,:,:,1) = res;

                % gradient and control filter update
                delta = zeros(obj.K,obj.J,obj.wlen);
                Xsf1 = permute(Xsf,[2 1 4 3]); % K x J x wlen x M
                for mm = 1:obj.M
                    delta = delta + obj.Err(mm,n) * Xsf1(:,:,:,mm);
                end
                obj.Wc = obj.Wc + muw * delta;
            end
    
        end


     end
end

File: python version/MultiChannelFxLMS.py
"""Fully connected J x K x M FxLMS translated from MATLAB.

The implementation intentionally retains the MATLAB dimension order:

- ``Wc``: K x J x Lw
- ``SecP``: M x K x Ls
- ``Ref``: J x N
- ``Dis``, ``y``, ``Err``: M x N
- ``yc``: K x N
"""

from __future__ import annotations

import numpy as np
from numpy.typing import ArrayLike, NDArray


class MultiChannelFxLMS:
    """Conventional fully connected multichannel FxLMS controller."""

    def __init__(
        self,
        wLen: int,
        Secondarypath: ArrayLike,
        sLen: int,
        ref: ArrayLike,
        dis: ArrayLike,
        numref: int,
        numsec: int,
        numerr: int,
    ) -> None:
        if min(wLen, sLen, numref, numsec, numerr) <= 0:
            raise ValueError(
                "wLen, sLen, numref, numsec, and numerr must be positive."
            )

        secondary_path = np.asarray(Secondarypath)
        reference = np.asarray(ref)
        disturbance = np.asarray(dis)
        dtype = np.result_type(
            secondary_path.dtype,
            reference.dtype,
            disturbance.dtype,
            np.float64,
        )

        self.J = int(numref)
        self.K = int(numsec)
        self.M = int(numerr)
        self.wlen = int(wLen)
        self.slen = int(sLen)
        self.SecP: NDArray = secondary_path.astype(dtype, copy=True)
        self.Ref: NDArray = reference.astype(dtype, copy=True)
        self.Dis: NDArray = disturbance.astype(dtype, copy=True)

        if self.Ref.ndim != 2 or self.Ref.shape[0] != self.J:
            raise ValueError(
                f"ref shape is {self.Ref.shape}; expected (J, N) with J={self.J}."
            )
        N = self.Ref.shape[1]
        expected_secondary_shape = (self.M, self.K, self.slen)
        expected_disturbance_shape = (self.M, N)
        if self.SecP.shape != expected_secondary_shape:
            raise ValueError(
                f"Secondarypath shape is {self.SecP.shape}; "
                f"expected {expected_secondary_shape} (M, K, Ls)."
            )
        if self.Dis.shape != expected_disturbance_shape:
            raise ValueError(
                f"dis shape is {self.Dis.shape}; "
                f"expected {expected_disturbance_shape} (M, N)."
            )

        self.Wc: NDArray = np.zeros(
            (self.K, self.J, self.wlen), dtype=dtype
        )
        self.yc: NDArray = np.zeros((self.K, N), dtype=dtype)
        self.y: NDArray = np.zeros((self.M, N), dtype=dtype)
        self.Err: NDArray = np.zeros((self.M, N), dtype=dtype)

    def McFxLMS_controller(self, muw: float) -> "MultiChannelFxLMS":
        """Run the fully connected FxLMS iteration."""

        N = self.Ref.shape[1]

        xc = np.zeros((self.J, self.wlen), dtype=self.Wc.dtype)
        ys = np.zeros((self.K, self.slen), dtype=self.Wc.dtype)
        xf = np.zeros((self.J, self.slen), dtype=self.Wc.dtype)
        Xsf = np.zeros(
            (self.J, self.K, self.M, self.wlen), dtype=self.Wc.dtype
        )

        for n in range(N):
            # MATLAB: xc = [Ref(:,n), xc(:,1:end-1)]
            xc[:, 1:] = xc[:, :-1]
            xc[:, 0] = self.Ref[:, n]

            # K control signals, each formed from all J references.
            for kk in range(self.K):
                self.yc[kk, n] = np.sum(self.Wc[kk, :, :] * xc)

            # M anti-noise signals through the M x K secondary paths.
            ys[:, 1:] = ys[:, :-1]
            ys[:, 0] = self.yc[:, n]
            for mm in range(self.M):
                self.y[mm, n] = np.sum(self.SecP[mm, :, :] * ys)

            self.Err[:, n] = self.Dis[:, n] - self.y[:, n]

            # Filter every reference through every M x K secondary path.
            xf[:, 1:] = xf[:, :-1]
            xf[:, 0] = self.Ref[:, n]
            res = np.zeros((self.J, self.K, self.M), dtype=self.Wc.dtype)
            for jj in range(self.J):
                for kk in range(self.K):
                    for mm in range(self.M):
                        res[jj, kk, mm] = np.dot(
                            xf[jj, :], self.SecP[mm, kk, :]
                        )

            Xsf[:, :, :, 1:] = Xsf[:, :, :, :-1]
            Xsf[:, :, :, 0] = res

            # MATLAB Xsf1 order: K x J x Lw x M.
            delta = np.zeros_like(self.Wc)
            for mm in range(self.M):
                delta += self.Err[mm, n] * np.transpose(
                    Xsf[:, :, mm, :], (1, 0, 2)
                )
            self.Wc += float(muw) * delta

        return self
